File: java/src/main/java/edu/ucdavis/fiehnlab/spectra/hash/core/types/SpectrumImpl.java
package edu.ucdavis.fiehnlab.spectra.hash.core.types;

import edu.ucdavis.fiehnlab.spectra.hash.core.Spectrum;

import java.util.*;

/**
 * simple implementation for a spectrum
 */
public class SpectrumImpl implements Spectrum {

    private String origin;
    private List<Ion> ions;
    private SpectraType type;

    public SpectrumImpl(List<Ion> ions,  String origin, SpectraType type) {
        this.ions = ions;
        this.origin = origin;
        this.type = type;
    }


    protected SpectrumImpl(){

    }

    public SpectrumImpl(List<Ion> ions, SpectraType ms) {
        this(ions,"unknown",ms);

    }

    public String getOrigin() {
        return origin;
    }

    public void setOrigin(String origin) {
        this.origin = origin;
    }

    public SpectraType getType() {
        return type;
    }

    public void setType(SpectraType type) {
        this.type = type;
    }

    public List<Ion> getIons() {
        return ions;
    }

    public void setIons(List<Ion> ions) {
        this.ions = ions;
    }


    public Spectrum toRelative(int scale) {
        double max = 0;

        for (Ion ion : getIons()) {
            if (ion.getIntensity() > max) {
                max = ion.getIntensity();
            }
        }

        ArrayList<Ion> ions = new ArrayList<Ion>(getIons().size());
        for (Ion ion : getIons()) {
            ions.add(new Ion(ion.getMass(), ion.getIntensity() / max * scale));
        }
        return new SpectrumImpl(ions, getOrigin(), getType());
    }

}

File: scala/src/main/scala/jp/riken/mirt/splash/scala/v1/SpectrumImpl.scala
package jp.riken.mirt.splash.v1

import edu.ucdavis.fiehnlab.spectra.hash.core
import jp.riken.mirt.splash.{ Ion, Spectrum }
import jp.riken.mirt.splash.JavaConversions._
import scala.collection.JavaConversions._

/**
 * Immutable wrapper for [[edu.ucdavis.fiehnlab.spectra.hash.core.types.SpectrumImpl]]
 */
trait SpectrumImpl extends core.types.SpectrumImpl with Spectrum 

/**
 * Factory object for the Java implementation of [[edu.ucdavis.fiehnlab.spectra.hash.core.types.SpectrumImpl]]
 */
object SpectrumImpl {
  def apply(ions: Seq[Ion], stype: core.types.SpectraType, origin: Option[String] = None): SpectrumImpl = {
    val _origin = origin.getOrElse("unknown")
    new core.types.SpectrumImpl(ions, _origin, stype) with SpectrumImpl
  }
}

